### ISSUE TITLE
Bug in channel implementation using mutex and condition variables

### ISSUE THREAD

**Issue by SumitPadhiyar** [ocaml-multicore/domainslib#25]
There seem to be a bug in channel implementation which causes [task_exn](https://github.com/ocaml-multicore/domainslib/blob/master/test/task_exn.ml) test to deadlock or loop infintely. The bug mainfest itself only in some runs otherwise it works fine.

**Comment by SumitPadhiyar** [comment]
I had debugged the test program and suspect that this has to do with channel implementation using single mutex+condvar for all channels. Exact root cause is when two receivers are waiting for a message [here](https://github.com/ocaml-multicore/domainslib/blob/3d5f79c5cf6a295b8c0652fb7470211b33b05b81/lib/chan.ml#L157) and when sender sends a message [here](https://github.com/ocaml-multicore/domainslib/blob/3d5f79c5cf6a295b8c0652fb7470211b33b05b81/lib/chan.ml#L88), it causes message to get lost. Receivers are waiting on same condition variable for the sender to write to a ref unique per waiting receiver. But the problem occurs when sender writes to ref r1(waited upon by receiver 1) and Condition_variable.signal signals wakes up receiver 2 which finds ref r2 not updated, so it waits again. So even after sending a meesage, both receiver are still waiting and the message is lost.

**Comment by ctk21** [comment]
Thanks for fuzzing domainslib, it is useful. 

I'm a bit confused how your proposed fix will work. Isn't there a unique mutex+condition_variable for each waiting receiver in `recv'` already:
https://github.com/ocaml-multicore/domainslib/blob/3d5f79c5cf6a295b8c0652fb7470211b33b05b81/lib/chan.ml#L150

I can see there is the problem if domainslib is used with systhreads.

**Comment by ctk21** [comment]
With systhreads and continuations/fibers it is possible for multiple execution contexts (aka 'threads') to share the same `Domain.DLS.get mutex_condvar_key` since `Domain.DLS` is specific at the _domain_ level.  

I make fixing the problem with systhreads and continuations/fibers being to replace `Condition.signal` with `Condition.broadcast`.

**Comment by SumitPadhiyar** [comment]
> With systhreads and continuations/fibers it is possible for multiple execution contexts (aka 'threads') to share the same `Domain.DLS.get mutex_condvar_key` since `Domain.DLS` is specific at the _domain_ level.
> 
> I make fixing the problem with systhreads and continuations/fibers being to replace `Condition.signal` with `Condition.broadcast`.

I agree on the fix of replacing  `Condition.signal` with `Condition.broadcast` as waiting receivers will still check for `ref Msg_slot` to be written, thereby waking up a single waiting receiver.
